bookstore/clone.py

```diff
@@ -406,3 +406,101 @@ def construct_template_params(self, relpath, fs_clonepath):
     def get_template(self, name):
         """Loads a Jinja template by name."""
         return BOOKSTORE_FILE_LOADER.load(self.settings['jinja2_env'], name)
+
+
+class BookstoreFSCloneAPIHandler(APIHandler):
+    """Handle notebook clone from a local file system.
+
+    Provides API handling for ``POST`` and clones a notebook
+    from the local file system.
+
+    Methods
+    -------
+    initialize(self)
+        Helper to access bookstore settings.
+    post(self)
+        Clone a notebook from the location specified by the payload.
+    build_content_model(self, content, path)
+        Helper for creating a Jupyter ContentsAPI compatible model.
+
+    See also
+    --------
+    `Jupyter Notebook reference on Custom Handlers <https://jupyter-notebook.readthedocs.io/en/stable/extending/handlers.html#registering-custom-handlers>`_
+    """
+
+    def initialize(self):
+        """Helper to retrieve bookstore setting for the session."""
+        self.bookstore_settings = BookstoreSettings(config=self.config)
+
+    def _get_content(self, path):
+        """Helper for getting content from a local filepath.
+        
+        Parameters
+        ----------
+        path : str
+            Source path from which to get a notebook
+        """
+        self.log.info(f"Reading content from {path}")
+        if os.path.splitext(path)[1] in [".ipynb", ".jpynb"]:
+            content = self.contents_manager._read_notebook(path)
+        else:
+            content = self.contents_manager._read_file(path, format=None)
+        return content
+
+    @web.authenticated
+    async def post(self):
+        """POST /api/bookstore/fs-clone
+
+        Clone a notebook to the path specified in the payload.
+
+        The payload type for the request should be::
+
+            {
+            "relpath": string,
+            "target_path"?: string
+            }
+
+        The response payload should match the standard Jupyter contents
+        API POST response.
+        """
+        model = self.get_json_body()
+
+        relpath = model.get("relpath", "")
+        fs_clonepath = validate_relpath(relpath, self.bookstore_settings, self.log)
+
+        target_path = model.get("target_path", "") or os.path.basename(os.path.relpath(relpath))
```

Comment:
If we are going to except a target path, we also need to make sure it is inside of the servers notebook directory. Or is that validation already done somewhere else?

Reply:
`self.contents_manager.save(content_model, content_model['path'])` has functionality that strips any prefixed `/` (making it effectively a relpath) and then requests to save it relative to the notebook's base directory.

That in turn runs `_get_os_path` on from the `filio.FileManagerMixin` to make sure that once it's resolved to the real path it is still kept underneath the basedir https://github.com/jupyter/notebook/blob/e498de6775b3de57f5ff827e562c179b17893064/notebook/services/contents/fileio.py#L243-L264